src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+{
+    private $connection;
+    private $tableItems;
+    private $tableVersions;
+
+    public function __construct(Connection $connection, string $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tableItems = "{$tablePrefix}_items";
+        $this->tableVersions = "{$tablePrefix}_versions";
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $this->connection->transactional(
+            function () use ($item) : void {
+                try {
+                    $current = $this->getCurrentVersion($item->getId(), LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    if (1 !== $item->getVersion()->toInt()) {
+                        throw new UnexpectedVersionNumber(
+                            $item->getId(),
+                            $item->getVersion(),
+                            ItemVersionNumber::fromInt(1)
+                        );
+                    }
+
+                    $this->insertNewItem($item);
+
+                    return;
+                }
+
+                $next = $current->next();
+
+                if ($item->getVersion() > $next) {
+                    throw new UnexpectedVersionNumber($item->getId(), $item->getVersion(), $next);
+                }
+
+                if ($item->getVersion() < $next) {
+                    $this->updateVersion($item);
+
+                    return;
+                }
+
+                $this->insertVersion($item);
+            }
+        );
+    }
+
+    public function remove(ItemId $id, ?ItemVersionNumber $version) : void
+    {
+        $this->connection->transactional(
+            function () use ($id, $version) : void {
+                if (null === $version) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                try {
+                    $current = $this->getCurrentVersion($id, LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    return;
+                }
+
+                if ($version > $current) {
+                    return;
+                }
+
+                if ($version < $current) {
+                    throw new UnexpectedVersionNumber($id, $version, $current);
+                }
+
+                if (1 === $current->toInt()) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                $this->deleteVersion($id, $current);
+            }
+        );
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (null === $version) {
+            $sql = 'SELECT version.id, version.version, version.content, version.hash'
+                ." FROM {$this->tableItems} item LEFT JOIN {$this->tableVersions} version"
+                .' ON item.id = version.id AND item.version = version.version'
+                .' WHERE item.id = :id';
```

Comment:
from this condition it seems this could return multiple rows, but only one is read by `fetch()`; any check that it is actually only one? Or a specified `ORDER BY` to default to the last/first version?

Reply:
`version.id` and `version.version` are the composite primary key.

Reply:
Ah, the constraint on `version` if I understand correctly only select 1 row here. Should be safe but due to bugs in joins it's generally worth to assert there are no results left after taking out the first result (Doctrine ORM used to do it in `findOne*()` methods)

Reply:
Falls in the what-happens-if-the-schema-is-not-right category. Will try and add error handling for these cases (and see if can add tests that hack the schema).